Index: docker.local/docker-compose-no-daemon.yml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/docker.local/docker-compose-no-daemon.yml b/docker.local/docker-compose-no-daemon.yml
new file mode 100644
--- /dev/null	(date 1635930579)
+++ b/docker.local/docker-compose-no-daemon.yml	(date 1635930579)
@@ -0,0 +1,42 @@
+version: "3"
+services:
+  mongodb:
+    image: mongo
+    volumes:
+      - ./0dns/mongodata:/data/db
+    networks:
+      default:
+      testnet0:
+    ports:
+      - "27017:27017"
+
+  0dns:
+    environment:
+      - DOCKER=true
+    build:
+      context: ./
+      dockerfile: ./docker.local/Dockerfile
+    depends_on:
+      - mongodb
+    links:
+      - mongodb:mongodb
+    volumes:
+      - ./config:/0dns/config
+      - ./0dns/log:/0dns/log
+    ports:
+      - "9091:9091"
+    networks:
+      default:
+      testnet0:
+        ipv4_address: 198.18.0.98
+    command: ./bin/zdns --deployment_mode 0 --magic_block /0dns/config/magic_block.json
+
+networks:
+  default:
+    driver: bridge
+  testnet0:
+    external: true
+
+volumes:
+  config:
+  mongodata:
Index: docker.local/docker-clean/docker-clean-compose.yml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/docker.local/docker-clean/docker-clean-compose.yml b/docker.local/docker-clean/docker-clean-compose.yml
new file mode 100644
--- /dev/null	(date 1635930579)
+++ b/docker.local/docker-clean/docker-clean-compose.yml	(date 1635930579)
@@ -0,0 +1,14 @@
+version: "3"
+services:
+  docker-clean:
+    build:
+      context: '.'
+      dockerfile: Dockerfile
+    volumes:
+      - ../0dns/mongodata:/data/db
+      - ../0dns/log:/0dns/log
+    command: /bin/sh docker-clean.sh
+
+volumes:
+  mongodata: {}
+  log: {}
Index: docker.local/docker-clean/docker-clean.sh
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/docker.local/docker-clean/docker-clean.sh b/docker.local/docker-clean/docker-clean.sh
new file mode 100644
--- /dev/null	(date 1635930579)
+++ b/docker.local/docker-clean/docker-clean.sh	(date 1635930579)
@@ -0,0 +1,10 @@
+#!/bin/sh
+
+set -e
+
+echo "cleaning up..."
+
+rm -rf /0dns/log/*
+rm -rf /data/db/*
+
+echo "cleaned up"
Index: docker.local/config/0dns.yaml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/docker.local/config/0dns.yaml b/docker.local/config/0dns.yaml
--- a/docker.local/config/0dns.yaml
+++ b/docker.local/config/0dns.yaml
@@ -1,8 +1,8 @@
 version: 1.0
 
 port: 9091
-use_https: true
-use_path: true
+use_https: false
+use_path: false
 
 logging:
   level: "info"
@@ -13,7 +13,7 @@ server_chain:
   signature_scheme: "bls0chain"
 
 handlers:
-  rate_limit: 5 # 5 per second
+  rate_limit: 1000 # 1000 per second
 
 worker:
   magic_block_worker: 5 # in seconds

Index: docker.local/docker-clean/Dockerfile
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/docker.local/docker-clean/Dockerfile b/docker.local/docker-clean/Dockerfile
new file mode 100644
--- /dev/null	(date 1635930579)
+++ b/docker.local/docker-clean/Dockerfile	(date 1635930579)
@@ -0,0 +1,7 @@
+
+#
+# clean up without sudo being a member of the docker group
+#
+
+FROM alpine:latest
+COPY docker-clean.sh ./docker-clean.sh
Index: docker.local/bin/start-no-daemon.sh
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/docker.local/bin/start-no-daemon.sh b/docker.local/bin/start-no-daemon.sh
new file mode 100755
--- /dev/null	(date 1635930579)
+++ b/docker.local/bin/start-no-daemon.sh	(date 1635930579)
@@ -0,0 +1,9 @@
+#!/bin/sh
+
+set -e
+
+PWD=`pwd`
+
+echo Starting 0dns ...
+
+docker-compose -p 0dns -f ./docker.local/docker-compose-no-daemon.yml up
Index: docker.local/bin/docker-clean.sh
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/docker.local/bin/docker-clean.sh b/docker.local/bin/docker-clean.sh
new file mode 100755
--- /dev/null	(date 1635930579)
+++ b/docker.local/bin/docker-clean.sh	(date 1635930579)
@@ -0,0 +1,11 @@
+#!/bin/sh
+
+#
+# clean up without sudo being a member of the docker group
+#
+
+set -e
+
+docker-compose                                                \
+    -f ./docker.local/docker-clean/docker-clean-compose.yml   \
+    up --build docker-clean
Index: docker.local/bin/stop.sh
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/docker.local/bin/stop.sh b/docker.local/bin/stop.sh
new file mode 100755
--- /dev/null	(date 1609674024602)
+++ b/docker.local/bin/stop.sh	(date 1609674024602)
@@ -0,0 +1,6 @@
+#!/bin/sh
+PWD=`pwd`
+
+echo Stopping 0dns ...
+
+docker-compose -p 0dns -f ../docker-compose.yml down
Index: docker.local/bin/stop-no-daemon.sh
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/docker.local/bin/stop-no-daemon.sh b/docker.local/bin/stop-no-daemon.sh
new file mode 100755
--- /dev/null	(date 1616847412304)
+++ b/docker.local/bin/stop-no-daemon.sh	(date 1616847412304)
@@ -0,0 +1,9 @@
+#!/bin/sh
+
+set -e
+
+PWD=`pwd`
+
+echo Starting 0dns ...
+
+docker-compose -p 0dns -f ./docker.local/docker-compose-no-daemon.yml down
